nilearn/decomposition/base.py

```diff
@@ -299,12 +299,16 @@ def fit(self, imgs, y=None, confounds=None):
             the affine is considered the same for all.
 
         """
-        if hasattr(imgs, '__iter__') and len(imgs) == 0:
+        if not hasattr(imgs, '__iter__'):
+            # these classes are meant for list of 4D images
+            # (multi-subject), we want it to work also on a single
+            # subject, so we hack it.
+            imgs = [imgs, ]
```

Comment:
I'm a bit afraid of this change as we don't really have control on the input. For example, with imgs="test.nii", this will not work as strings are iterable. And I'm sure that we can find other weird cases. This is why NiftiStuffMasker are declined in Single and Multi flavors: accepting any dimensionality of data was too messy.

Plus documentation is not updated.

I would go for feeding ['test.nii'] to canica for single subject and postponing this behavior for later when we will have time to think it through. Apart from that, +1 for merge.